### PR TITLE
nixos/systemd-tmpfiles: add settings option

### PR DESCRIPTION
**Pull request by Lassulus**
## Description of changes

Add a more declarative option to create tmpfiles so people can read/write those without having the manual open all the time. Suggestions how to make it even cooler are welcome.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/system/boot/systemd/tmpfiles.nix
Comment (arianvp) [inline nixos/modules/system/boot/systemd/tmpfiles.nix:165]: This should probably be `10-nixos-${namespace}.conf` to make sure it runs first?

or `99` for last. we need to make some decision here.

Comment (Lassulus) [reply nixos/modules/system/boot/systemd/tmpfiles.nix:165]: both seem fine for me? we have also a bunch of default ones in there which don't have a prefix number

File: nixos/modules/system/boot/systemd/tmpfiles.nix
Comment (aanderse) [inline nixos/modules/system/boot/systemd/tmpfiles.nix:70]: i feel like we can leave a better example... maybe `"10d`"?

File: nixos/modules/system/boot/systemd/tmpfiles.nix
Comment (ElvishJerricco) [inline nixos/modules/system/boot/systemd/tmpfiles.nix:167]: This would be better as a `writeTextDir` rather than a `runCommand`.

And I'd prefer to see it written with `mapAttrsToList` rather than using `attrNames` and mapping with `foo.${name}` syntax.

File: nixos/modules/system/boot/systemd/tmpfiles.nix
Comment (aanderse) [inline nixos/modules/system/boot/systemd/tmpfiles.nix:24]: `description` values should follow proper sentence structure and grammar

additionally... if we're going to the effort of making a really nice wrapper like this (❤️ @Lassulus) we should really make a good description that explains why NixOS has a special relationship with `tmpfiles` - our love of declarative approach over `bash` scripts - and specifically emphasize that `tmpfiles` is a misleading name

i mention this because the name really throws off new people! outside of NixOS this feature of `systemd` is not as popular and well known so we should really clarify the situation... then people can say "that nixos distro has really good docs!" 😄

File: nixos/modules/system/boot/systemd/tmpfiles.nix
Comment (aanderse) [reply nixos/modules/system/boot/systemd/tmpfiles.nix:163]: since we're using attribute sets we should really consider putting our options higher priority - ready to override because of module merging, more power

File: nixos/modules/system/boot/systemd/tmpfiles.nix
Comment (aanderse) [inline nixos/modules/system/boot/systemd/tmpfiles.nix:163]: if we're leaving quoting of arguments to the user we should explicitly mention this... or just take care of it for them

**Review by aanderse — DISMISSED** [review]
awesome stuff!

File: nixos/modules/system/boot/systemd/tmpfiles.nix
Comment (ElvishJerricco) [inline nixos/modules/system/boot/systemd/tmpfiles.nix:36]: I think this is one `attrsOf` too deep. I'd rather write `type = "d";` than use an `attrsOf` for that last level, particularly because it currently indicates you should be able to do this:

```
systemd.tmpfiles.settings.foo."/bar" = {
  d.mode = "0755";
  f.argument = "baz";
};
```

Of course this is nonsense, since one path can only have one entry, but here I've made that path both a directory and a file.

**Review by ElvishJerricco — DISMISSED** [review]

File: nixos/modules/system/boot/systemd/tmpfiles.nix
Comment (aanderse) [reply nixos/modules/system/boot/systemd/tmpfiles.nix:36]: > since one path can only have one entry,

you can have multiple entries per path

some types like `f` and `d` conflict but others do not
we need the ability to specify multiple - this is used in nixos and beyond

Comment (Lassulus) [reply nixos/modules/system/boot/systemd/tmpfiles.nix:36]: it would be nice to assert those types that can conflict and which not, but I have no good idea how to do that.

Comment (ElvishJerricco) [reply nixos/modules/system/boot/systemd/tmpfiles.nix:36]: I stand corrected :)

File: nixos/modules/system/boot/systemd/tmpfiles.nix
Comment (Lassulus) [reply nixos/modules/system/boot/systemd/tmpfiles.nix:24]: I wanted to extend the description after we are happy with the interface :)

File: nixos/modules/system/boot/systemd/tmpfiles.nix
Comment (Lassulus) [reply nixos/modules/system/boot/systemd/tmpfiles.nix:163]: how to properly quote them? can we wrap them in " quotes?

File: nixos/modules/system/boot/systemd/tmpfiles.nix
Comment (Lassulus) [reply nixos/modules/system/boot/systemd/tmpfiles.nix:167]: I can try to rework it to that. basically I just copied the code from above and extended it a bit until it worked.

Comment (Lassulus) [reply nixos/modules/system/boot/systemd/tmpfiles.nix:167]: hmm, it seems writeTextDir creates only a single file. so we would have a derivation for each "namespace"/file. That doesn't seem like a lot of overhead but could be slower if we have a lot of namespaces

Comment (Lassulus) [reply nixos/modules/system/boot/systemd/tmpfiles.nix]: I changed it now to use writeTextDir and mapAttrsToList

File: nixos/modules/system/boot/systemd/tmpfiles.nix
Comment (Lassulus) [reply nixos/modules/system/boot/systemd/tmpfiles.nix]: seems that it works for my simple test case with single quotes

File: nixos/modules/system/boot/systemd/tmpfiles.nix
Comment (Lassulus) [reply nixos/modules/system/boot/systemd/tmpfiles.nix:36]: at what point should the check run? I guess if we use writeTextDir we can't really do any verification on the file?

Comment (ElvishJerricco) [reply nixos/modules/system/boot/systemd/tmpfiles.nix:36]: That's fair, though I'm not sure systemd actually *has* a validation command for tmpfiles. But if it does and we have to use `runCommand` to validate with it, we should still *definitely* use `passAsFile` rather than shell-escaping arguments to `echo`.

File: nixos/modules/system/boot/systemd/tmpfiles.nix
Comment (aanderse) [reply nixos/modules/system/boot/systemd/tmpfiles.nix]: https://www.freedesktop.org/software/systemd/man/latest/tmpfiles.d.html

> Fields may contain C-style escapes. With the exception of the seventh field (the "argument") all fields may be enclosed in quotes. Note that any whitespace found in the line after the beginning of the argument field will be considered part of the argument field. To begin the argument field with a whitespace character, use C-style escapes (e.g. "\x20").

ok so we should quote all fields except the last one, `argument`

File: nixos/modules/system/boot/systemd/tmpfiles.nix
Comment (nikstur) [reply nixos/modules/system/boot/systemd/tmpfiles.nix:37]: There is no validation command for systemd-tmpfiles. I think it's fine to not add a validation derivation here, we can always add a one later.

File: nixos/modules/system/boot/systemd/tmpfiles.nix
Comment (nikstur) [reply nixos/modules/system/boot/systemd/tmpfiles.nix]: We should honestly leave this to the user, so `"${namespace}.conf"`. The the user can decide for themselves whether they need priority over other settings or not. This is also in line with what we do with many other systemd tools in Nixpkgs.

File: nixos/modules/system/boot/systemd/tmpfiles.nix
Comment (nikstur) [reply nixos/modules/system/boot/systemd/tmpfiles.nix]: Just extended it (with @lassulus permission)!

Comment (nikstur) [reply nixos/modules/system/boot/systemd/tmpfiles.nix]: Just extended it (with @lassulus permission)!

**Comment by aanderse** [comment]
how does everyone feel about having to specify a filename? is this easy for users to understand? would it be useful to put a priority inside the options itself instead of having to specify a filename? usually i don't care about filename so this could get tedious

... or not? any opinions?

**Comment by nikstur** [comment]
> how does everyone feel about having to specify a filename? is this easy for users to understand? would it be useful to put a priority inside the options itself instead of having to specify a filename? usually i don't care about filename so this could get tedious
> 
> ... or not? any opinions?

I have quite a strong opinion about this. We already have a very similar API in other systemd services (at least sysupdate and repart) where you have to specify the filename. Usually you don't have to care about the filename (as you often don't need ordering) but when you want to, or need to, you can with this API. Generally we try to stay close to the way upstream systemd works and if you read the systemd manpages (which you have to if you want to use tmpfiles) this API should make a lot of sense.

**Comment by aanderse** [comment]
> if you read the systemd manpages (which you have to if you want to use tmpfiles) this API should make a lot of sense.

it was explicitly stated that the entire purpose of this PR was to spare people having to read the manpage

this PR is meant to be an abstraction where the other systemd pieces you mention are not

I'm not saying we shouldn't do we're you're saying... but i am saying your reasoning mentioned here isn't flowing smoothly with me

🤔

**Comment by nikstur** [comment]
> it was explicitly stated that the entire purpose of this PR was to spare people having to read the manpage

You will for sure have to read the manual, just to understand the types of operations---but you don't have to read it *all the time* because the basic syntax (e.g. sets) is convenient and familiar as opposed to random strings. There is no way around reading the manual (and it also shouldn't be our goal to achieve independence of upstream documentation).

> this PR is meant to be an abstraction where the other systemd pieces you mention are not

The other pieces are abstractions as well. They are small wrappers (i.e. structural settings) around the systemd API---the same as what we are proposing here.

**Comment by aanderse** [comment]
if that's the intention I'll take that as a satisfactory answer then
ok, thanks again everyone for working on this

**Review by nikstur — APPROVED** [review]

**Comment by arianvp** [comment]
> ... or not? any opinions?

for `systemd.tmpfiles.rules` we also don't define the filename. It gets written to `00-nixos.conf` instead. Bit confused that this wasn't doing that.

For things like `.network` files we can't do something similar as `matchConfig` is fuzzy and there is no clear way of merging them. But for `tmpfiles` I'd much rather prefer using the NixOS module system for merging .

But it seems @nikstur and I have opposing opinions on this. Perhaps we can discuss this a bit further?

I am probably easily convinced that my opinion is wrong. It's just confusing that `tmpfiles.rules` doesn't work like this ?

**Comment by aanderse** [comment]
> I am probably easily convinced that my opinion is wrong. It's just confusing that `tmpfiles.rules` doesn't work like this ?

that is sorta where i was at :man_shrugging:

> Perhaps we can discuss this a bit further?

if i'm not the only person who felt this way then i would like that...

**Comment by nikstur** [comment]
> It's just confusing that tmpfiles.rules doesn't work like this?

I agree that it's confusing that the "old" tmpfiles did not work like that. This was actually one of the reasons to prompt me to push for this design (where you have control over the filenames) to make it more compatible with the other systemd abstractions we provide.

To give some additional context: @DavHau , @Lassulus and me worked on this in person and discussed and iterated on the API.

**Comment by aanderse** [comment]
> To give some additional context: @DavHau , @Lassulus and me worked on this in person and discussed and iterated on the API.

and you all thought merging in `nix` wasn't desirable here? i'm thinking about this from the perspective of someone who isn't familiar with `tmpfiles` (read: almost everyone outside of the nixos world? `tmpfiles` has not caught on in other distros or upstream to the extent it has with us, and even with us you'll see many people get confused (hence why this PR is a good idea) ) and i keep coming back to "simple is good"

i guess we're just disagreeing on what we use as examples to justify "consistent"... i see your point, i see my point... both are valid :thinking:

**Comment by Lassulus** [comment]
another upside of the current implementation is the easier tracking why a rule exists. with .rules every declaration ended up in one big file. now it's easier to match the generated files to the module which actually generates them?

**Comment by RaitoBezarius** [comment]
Personally, in this context, there's either perfect or good enough.

Perfect should probably take the form of using the convergence script to connect to some varlink API or similar to inject the merged configuration options with provenance and we can wire it up in the NixOS modules and use the mkPriority stuff.

Good enough is letting users do informal priority via the lexorder and it's sad because no introspection at Nix level but we recover provenance on the disk and it's nice for debugging.

Usually, anyway, it's already very hard to introspect the existing rules because they are all over the places in a NixOS system. Provenance seems appealing for me and this PR would improve the situation.

I understand how frustrating this can be from an conceptual perspective, but I believe that operational considerations would trump things here for me.

Of course, we can revisit later if the API is a PITA, let's not make the choice of something that locks us into something.

**Review by RaitoBezarius — APPROVED** [review]

**Comment by Lassulus** [comment]
@aanderse you still have a negative review which is more or less blocking the merge

**Comment by aanderse** [comment]
i think the discussion was good enough that we should merge
i still hold my opinion... but i see your points, so let's agree to disagree

regardless... in my opinion this interface is definitely better than the current, thanks for doing this!

**Comment by SuperSandro2000** [comment]
I've just noticed this and I must say this is a really nice improvement. Writing the rules before was always awkward.

**Comment by arianvp** [comment]
Happy to hear!

**Pull request merged**
